### PR TITLE
Ensure tokenize is thread safe

### PR DESCRIPTION
**Pull request by fjetter**
I ran into some serious problem with this over in https://github.com/dask/dask/pull/11248 because some tasks are now tokenized differently. Particularly, SubgraphCallable functions are no longer pickled but break down into proper tokens. Particularly the section that handles cyclic references in sequences cannot deal with this very well. My solution to this is to just slap a lock on this. tokenize is only called very rarely in a thread so I don't anticipate much congestion.

Note that there is still a lingering determinism problem I had to dedent a couple of test classes because they would take the cloudpickle path which, despite being locked, still produce weird results intermittently. I still think this is an improvement.

### CODE REVIEW

File: dask/tests/test_tokenize.py [hunk 1]
```diff
@@ -50,7 +52,44 @@ def check_contextvars():
...
+
+
+def _init_tpe() -> ThreadPoolExecutor:
+    pytest.importorskip("psutil")
+    global _TPE
+    if _TPE is not None:
+        return _TPE  # type: ignore[unreachable]
+    cores = psutil.cpu_count(logical=False)
+    _TPE = ThreadPoolExecutor(cores * 4)
+    return _TPE
```
Comment (fjetter) [inline dask/tests/test_tokenize.py:74]: This is a bit of a hack to avoid having to pass this through every function... hope nobody minds 🙈

**Comment by phofl** [comment]
The 2 test failure are tokenise related, so that seems legit?

**Comment by fjetter** [comment]
yes, I'm looking into this further. The failures are not related to threadsafety but are indeterministic tokenizations that are caused by cloudpickle. The cloudpickle output is not deterministic. The streams I'm looking at, something is messing with the pickle memoization such that pickle streams end up different... what a mess!__

**Comment by fjetter** [comment]
I believe by now that this is in fact not a thread safety issue but some cloudpickle (or possibly even plain pickle) internal problem. I tried to narrow this down but ultimately failed.

From what I can currently tell, the unpickling of cloudpickle is not identifying some objects as identical and is recreating them which causes follow up pickles to be memoizing incorrectly, therefore changing the bytestream and therefore the token... 

I'm abandoning this PR. I think I can work around it on my other PR but this is generally a bit concerning

**Pull request closed without merging**